File: code/go/ch15/src/luago/state/closure.go
package state

import . "luago/api"
import "luago/binchunk"

type upvalue struct {
	val *luaValue
}

type closure struct {
	proto  *binchunk.Prototype // lua closure
	goFunc GoFunction          // go closure
	upvals []*upvalue
}

func newLuaClosure(proto *binchunk.Prototype) *closure {
	upvals := make([]*upvalue, len(proto.Upvalues))
	return &closure{
		proto:  proto,
		upvals: upvals,
	}
}

func newGoClosure(f GoFunction, nUpvals int) *closure {
	upvals := make([]*upvalue, nUpvals)
	return &closure{
		goFunc: f,
		upvals: upvals,
	}
}

File: code/csharp/ch10/luavm/state/Closure.cs
using luavm.api;
using luavm.binchunk;

namespace luavm.state
{
    public struct Upvalue
    {
        internal object val;
    }

    public class Closure
    {
        internal Prototype proto;
        internal CsharpFunction csharpFunc;
        internal Upvalue[] upvals;

        internal static Closure newLuaClosure(Prototype proto)
        {
            var c = new Closure
            {
                proto = proto
            };
            if (proto.Upvalues != null && proto.Upvalues.Length > 0)
            {
                c.upvals = new Upvalue[proto.Upvalues.Length];
            }

            return c;
        }

        internal static Closure newCsharpClosure(CsharpFunction charpFunc, int nUpvals)
        {
            var c = new Closure
            {
                csharpFunc = charpFunc
            };
            if (nUpvals > 0)
            {
                c.upvals = new Upvalue[nUpvals];
            }

            return c;
        }
    }
}